### PR TITLE
Add edpm_reboot_strategy to reboot role

### PR DESCRIPTION
**Pull request by ciecierski**
Add edpm_reboot_strategy to reboot role that defines if reboot should be started or not. There are three modes: auto, never and force. The default is auto.

Add also more generic check for existence of openstack configs for already
deployed nodes, both old tripleo and new edpm node.

### CODE REVIEW

File: roles/edpm_reboot/tasks/main.yaml
Comment (slagle) [inline roles/edpm_reboot/tasks/main.yaml:69]: the already deployed check only considers the existence of nova.conf, so perhaps that should be the description. either that, or we need to add a way to flag an edpm node as already deployed independent of nova as there could be other node types.

Comment (ciecierski) [reply roles/edpm_reboot/tasks/main.yaml:69]: I discussed it with Jiri and he mentioned networker nodes with ovn. I'm going to add check for ovn files on nodes to include networker nodes. Do you know other node types and check for those?

File: roles/edpm_reboot/defaults/main.yaml
Comment (jistr) [inline roles/edpm_reboot/defaults/main.yaml:21]: IIUC this should match also on legacy compute nodes, which is nice, since the nova.conf approach probably will not match those (as the config path is different in 17.1).

**Review by jistr — COMMENTED** [review]
LGTM, thanks :+1:

File: roles/edpm_reboot/tasks/main.yaml
Comment (ciecierski) [reply roles/edpm_reboot/tasks/main.yaml]: @slagle In last revision I made a change that makes the check independent of nova conf. Please check pull request once again

**Review by bogdando — COMMENTED** [review]
lgtm

File: roles/edpm_reboot/tasks/main.yaml
Comment (gibizer) [inline roles/edpm_reboot/tasks/main.yaml:82]: does it need to check not just the old but the new nova conf existence?

Comment (ciecierski) [reply roles/edpm_reboot/tasks/main.yaml:82]: It does check in next line.83 (and not edpm_node_conf.stat.exists)
Strictly speaking it checks for any openstack conf that may exists in config paths, because there was an ask to make check more node type agnostic.

Comment (ciecierski) [reply roles/edpm_reboot/tasks/main.yaml:82]: No problem, thank you for reviewing change.

**Comment by jistr** [comment]
What is Tide waiting for here in terms of approval? Does it need to be approved by @SeanMooney specifically as the comment above suggests? :)

**Comment by jistr** [comment]
Ah i guess Mikolaj or i would have to be on the approvers list.

**Comment by SeanMooney** [comment]
no it just requries approval form anyone in the owners file
https://github.com/openstack-k8s-operators/edpm-ansible/blob/main/OWNERS
or OWNERS_ALIASES
https://github.com/openstack-k8s-operators/edpm-ansible/blob/main/OWNERS_ALIASES

File: roles/edpm_bootstrap/tasks/fips.yml
Comment (SeanMooney) [inline roles/edpm_bootstrap/tasks/fips.yml:64]: are we sure that a forced rebbot is something well should
enable by default in any of our roles?

File: roles/edpm_reboot/defaults/main.yaml
Comment (SeanMooney) [inline roles/edpm_reboot/defaults/main.yaml:21]: im not sure why this was ever using /var/lib/openstack/config/nova

that was obviously a bug because that directory si reserved for use by the edpm_nova role

we will delete anything in that and recreated every time it runs.

so +1 but we may want to add a linint tests for this at some point to ensure other roles to try to modify files in /var/lib/openstack/config/nova

File: roles/edpm_reboot/tasks/main.yaml
Comment (SeanMooney) [inline roles/edpm_reboot/tasks/main.yaml:59]: ok so this is trying to skip reboot if there is tripleo content.

we don't every delete those config files during adoption so this will mean we will never reboot. compute nodes in auto mode for any adopted nodes.

**Review by SeanMooney — CHANGES_REQUESTED** [review]
so this pr looks mostly ok but i has no test coverage.
ideally this would have  molecule coverage although I'm not entirely sure how to test this properly.
the reason i marked this as request change is two fold.

first is there a generic adoption task planned to clean up tripleo config? by default we likely should be avoiding remoood for edpm nodes with nova but I'm not sure its safe to force the reboot for fpis mode by default.

so we should likely see what other think before merging

File: roles/edpm_reboot/tasks/main.yaml
Comment (SeanMooney) [reply roles/edpm_reboot/tasks/main.yaml:59]: if there is a jira to plan a post adoption task to clean up these stale config i think we can also ignore this point and proceed with the pr

File: roles/edpm_bootstrap/tasks/fips.yml
Comment (ciecierski) [reply roles/edpm_bootstrap/tasks/fips.yml:64]: It was 'forced'  before I changed it to edpm_reboot_force_reboot and now to  edpm_reboot_strategy: force . If this is wrong, this is existing bug in tripleo too.

File: roles/edpm_reboot/defaults/main.yaml
Comment (ciecierski) [reply roles/edpm_reboot/defaults/main.yaml:21]: well, in first iteration we took care of computes only. Later it was brought to attention that there are other node types we should check and not reboot.

File: roles/edpm_reboot/tasks/main.yaml
Comment (ciecierski) [reply roles/edpm_reboot/tasks/main.yaml:59]: Auto mode is there only for greenfield deployment only, so regardless if cleanup happen(it should happen I agree) adopted nodes won't be rebooted in auto mode anyway(after adoption they will have new configs generated).

**Comment by ciecierski** [comment]
If there are concerns how this change works for adoption I link change that tests the change in pre-adoption phase.
https://github.com/openstack-k8s-operators/data-plane-adoption/pull/459/files

**Comment by SeanMooney** [comment]
So that pr is not depending on this pr so it's not testing it and it's also failing ci which is not very encouraging 

I'll remove the requested change to allow others to approve but I'm still unsure that how reboot are being done are correct

File: roles/edpm_reboot/tasks/main.yaml
Comment (SeanMooney) [reply roles/edpm_reboot/tasks/main.yaml:59]: So that is problematic because after a deployment is adopted it is ment to be identical to a greenfield deployment with regards to future day 2 operation. So post adoption completing auto should work the same for greenfield and adopted node otherwise we have a big.

Comment (bogdando) [reply roles/edpm_reboot/tasks/main.yaml:59]: that is correct, it seems that users then would need to change all osdpns ansible vars for that policy to auto and / or use this value for the future osdp deployments, on "day 2", post-adoption. That is a docs impact though, nothing for this code to change

**Comment by SeanMooney** [comment]
the os-reboot will be used for kernel updates or other changes that require a host reboot as part of day two maintenance.

it feels wrong to require force to be used in that case
auto should work asuimg a prior task has marked the node as needing reboot.

the intended workflow is that you would do a separate deployment including the os-reboot service with --limit to reboot subsets of host that you have drained of workloads in batches.

in that workflow we should not need to update the ansible_vars on the nodeset to force.

if we cant support that usecase we cant do minor updates properly.

**Review by SeanMooney — APPROVED** [review]

**Comment by SeanMooney** [comment]
i will file a jira bug for the fips issue

**Comment by ciecierski** [comment]
> So that pr is not depending on this pr so it's not testing it and it's also failing ci which is not very encouraging
Yes, the pr wasn't testing it. I made a patch, but there was missing depends-on. I fixed it and new results are https://github.com/openstack-k8s-operators/data-plane-adoption/pull/459#issuecomment-2133562264 
A non-ceph job failed during adoption, it passed pre-adoption validation with reboot.

**Comment by SeanMooney** [comment]
filed [OSPRH-7301](https://issues.redhat.com/browse/OSPRH-7301)

**Pull request merged**
